### PR TITLE
Imagery Offset Database

### LINKED ISSUE #1124: Use the upcoming imagery offset database
Manually adjusting every layer is hard, non-obvoius and usually unneeded task which duplicates the efforts of previous mappers in an area. In a week or two I'm announcing the imagery offset server which allows users to determine an offset for each area once, and then just use the shared value.

See [the description](http://wiki.openstreetmap.org/wiki/Imagery_Offset_Database) and [API documentation](http://wiki.openstreetmap.org/wiki/Imagery_Offset_Database/API). There is also [a map](http://offsets.textual.ru/map) and [a list](http://offsets.textual.ru/all) (refreshes every hour).

Also I'd like you to check [the alpha-version of a JOSM plugin](https://www.dropbox.com/s/abqlwa57pkrck81/imagery_offset_db.jar) (it should be put into `~/.josm/plugins`). It shows that the user interface mostly consists of informational messages and a dialog for choosing an offset. The latter is the most important part: its design should give users enough information to correctly choose an offset and rid them of doing the same work twice.

I plan to release beta on Monday or Tuesday, and before that I'm open to any suggestions, even those which could reqiuire structural changes to the database and the plugin.

### PR DESCRIPTION
**Pull request by kepta**
@bhousel would need help in figuring out certain things.

- What do we keep the padding of imagery points ?
- Where should we invoke the check for imagery offset ? Should it happen automatically or be shown in the imagery adjustment UI, like click to apply saved imagery offset.

(closes #1124 )

### CODE REVIEW

**Comment by bhousel** [comment]
This is a great start - just a reminder to use [4 space indent](https://github.com/openstreetmap/iD/blob/master/CONTRIBUTING.md#javascript), like we do everywhere else in iD.  (we should probably have an [`.editorconfig`](http://editorconfig.org/) file in our project root that enforces this.)

> What do we keep the padding of imagery points ?

Ideally we don't want to pad the lookup and store in a location cache - it would be better to cache the results of each tile url that we lookup.  I'm not sure whether the service could handle the load of all iD users asking for offset for every tile.  What do you think @Zverik ?

> Where should we invoke the check for imagery offset ? Should it happen automatically or be shown in the imagery adjustment UI, like click to apply saved imagery offset.

Ideally automatically.  We store the [user supplied offset here](https://github.com/openstreetmap/iD/blob/master/modules/renderer/background_source.js#L29) in `background_source`.  We could add a bool variable there to say whether the offset is automatically fetched, and we could add a corresponding checkbox on the Background ui screen.

**Comment by bhousel** [comment]
@kepta As a next step, can you add a test here:  https://github.com/openstreetmap/iD/tree/master/test/spec/services

**Comment by Zverik** [comment]
Thanks for the pull request. Automatically checking the offset for every tile is an overkill and will quickly put the server offline. I suggest doing that when opening the editor, and after that when a user scrolls the map more than a kilometer (at least 500 m) off the last offset checking point.

**Comment by kepta** [comment]
I have added some simple tests.
Next actions would be, to change the way search works. Currently the backend gives the whole list of imagery's without the provided radius. We would like to cache all of it but return back only the imagery offset matching with the active imagery. 
The backend gives the data_url of the imagery, I guess that can be used to map the active imagery to the one in the cache?.

**Comment by bhousel** [comment]
> My bad forgot to check the indentation, I have fixed it in the latest commit. @bhousel should I rebase and squash the commits, to make it look better?

No problem.  I don't care much about how the commit history looks.

> I have added some simple tests and also added a padding of 1km to each imagery pos as per @Zverik 's recommendation. I am still not sure about the radius to search in.

You could probably use a large radius, since we're caching the results  (so that would mean fewer calls) - and the results packets are small, and there probably aren't too many.

I guess try it with a few different search radii and see?

**Comment by kepta** [comment]
> I guess try it with a few different search radii and see?
That makes sense, I will go for an increased radius. Any thoughts on how to match the active imagery with the one in cache?

**Comment by bhousel** [comment]
> That makes sense, I will go for an increased radius. @bhousel Any thoughts on how to match the active imagery with the one in cache?

Yes, I think we can do this similarly to how I sample [imagery vintage in the Background panel](https://github.com/openstreetmap/iD/blob/master/modules/ui/panels/background.js#L65-L81).

The background layer is a bunch of `` elements, and one of them which is closest to the center of the screen will be classed with `.tile-center`.

Here is some code that you can stuff into [the background panel](https://github.com/openstreetmap/iD/blob/master/modules/ui/panels/background.js).  There is already a `getVintage` function, so this would be how you might write a `getOffset` function.

```js
var tile = d3.select('.layer-background img.tile-center');   // tile near viewport center
if (tile.empty()) return;

var d = tile.datum();   // the data for that tile
if (!d || d.length < 4) return;

var zoom = d[2];
var url = d[3];
var center = context.map().center();

// service.getOffset will work somewhat like our nominatim service reverse():
// 1. use magic to turn the url, center, and zoom level into an identifier, per
//     https://wiki.openstreetmap.org/wiki/Imagery_Offset_Database/API#TMS
// 2. query the offset database with those parameters (or return cached result)
// 3. cache result and call the callback
service.getOffset(url, center, zoom, function(err, result) {
   // console.log something?
   // or update a div someplace with the numbers
});
```

File: modules/services/imagery_offset.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    },
+
+    reset: function() {
+        _.forEach(inflight, function(req) {
+            req.abort();
+        });
+        inflight = {};
+        offsetCache = rbush();
+    },
+    getImageryID: function(url, imageryType) {
```
Comment (kepta) [inline modules/services/imagery_offset.js]: @Zverik could I get your review for this piece of code, I tried to convert the Java [code](https://trac.openstreetmap.org/browser/subversion/applications/editors/josm/plugins/imagery_offset_db/src/iodb/ImageryIdGenerator.java) you wrote in Javascript.

If you could also point me to some unit test cases for this function in your code base ..

**Comment by kepta** [comment]
@bhousel do we support `WMS` type of imageries in iD, since I couldn't find one in `imagery.json`?

**Comment by bhousel** [comment]
> do we support WMS type of imageries in iD, since I couldn't find one in imagery.json?

Not yet.. that's #1141

File: modules/services/imagery_offset.js [hunk 2]
```diff
@@ -22,8 +22,95 @@ export default {
...
+        if (url == null) return;
+        console.log('url=', url)
+        debugger;
+        url = url.toLowerCase();
+        if (imageryType === 'bing' || url.indexOf('tiles.virtualearth.net') > -1)
+            return 'bing';
+        // if (url.indexOf("scanex_irs") > )
+        if (
+            imageryType === 'TMS' &&
+            url.match(/.+tiles\.mapbox\.com\/v[3-9]\/openstreetmap\.map.*/)
```
Comment (pnorman) [inline modules/services/imagery_offset.js]: A sed pattern (`s/foo/bar/`) can use any character to delimit, and it's normal to use something other than `/` when dealing with URLs, e.g. `s@foo@bar@`. This avoids excessive escaping. Is this possible with Javascript's regex implementation?

File: modules/ui/background.js [hunk 3]
```diff
@@ -185,7 +185,7 @@ export function uiBackground(context) {
         function update() {
             backgroundList.call(drawList, 'radio', clickSetSource, function(d) { return !d.overlay; });
             overlayList.call(drawList, 'checkbox', clickSetOverlay, function(d) { return d.overlay; });
-
+
```
Comment (pnorman) [inline modules/ui/background.js]: Stray whitespace

File: modules/ui/panels/background.js [hunk 4]
```diff
@@ -81,6 +86,35 @@ export function uiPanelBackground(context) {
...
+        var tile = d3.select('.layer-background img.tile-center'); // tile near viewport center
+        if (tile.empty()) return;
+
+        var d = tile.datum();
+        var zoom = d[2];
+        var url = d[3];
+        var center = context.map().center();
+        console.log(d);
+        searchOffset.search(center,url, console.log);
+        //     zoom = (d && d.length >= 3 && d[2]) || Math.floor(context
```
Comment (pnorman) [inline modules/ui/panels/background.js]: Should this commented code be removed?

File: modules/ui/panels/background.js [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
+var searchOffset = services.imageryOffset;
+searchOffset.init();
 
 export function uiPanelBackground(context) {
     var background = context.background();
     var currSource = null;
     var currZoom = '';
     var currVintage = '';
-
+    debugger;
```
Comment (pnorman) [inline modules/ui/panels/background.js]: This seems out of place

File: modules/services/imagery_offset.js [hunk 6]
```diff
@@ -22,8 +22,95 @@ export default {
...
+            // TMS: skip parameters with variable values and Mapbox's access token
+            if (
+                (kv.length > 1 &&
+                    kv[1].indexOf('{') >= 0 &&
+                    kv[1].indexOf('}') > 0) ||
+                kv[0] === 'access_token'
+            ) {
+                return;
+            }
+            console.log('here')
```
Comment (pnorman) [inline modules/services/imagery_offset.js]: Left over debugging statement?

File: modules/services/imagery_offset.js [hunk 7]
```diff
@@ -22,8 +22,95 @@ export default {
...
+        ];
+
+        var qparams = {};
+        var qparamsStr = query.length > 1 ? query.slice(1).split('&') : '';
+
+        qparamsStr.forEach(function(param) {
+            var kv = param.split('=');
+            console.log(kv);
+            kv[0] = kv[0].toLowerCase();
+            // WMS: if this is WMS, remove all parameters except map and layers
```
Comment (pnorman) [inline modules/services/imagery_offset.js]: Not needed for iD

File: modules/services/imagery_offset.js [hunk 8]
```diff
@@ -22,8 +22,95 @@ export default {
...
+
+        var query = '';
+        // Split URL into address and query string
+        var questionMarkIndex = url.indexOf('?');
+        if (questionMarkIndex > 0) {
+            query = url.slice(questionMarkIndex);
+            url = url.slice(0, questionMarkIndex);
+        }
+
+        var removeWMSParams = [
```
Comment (pnorman) [inline modules/services/imagery_offset.js]: Not needed for iD

File: modules/services/imagery_offset.js (hunk 1 above)
Comment (pnorman) [inline modules/services/imagery_offset.js]: JavaScript probably contains more (or better) files for handling URLs. I'm wondering if it makes sense to write this differently, particularly when parsing the parameters part of the URL

**Comment by pnorman** [comment]
> > do we support WMS type of imageries in iD, since I couldn't find one in imagery.json?
> Not yet.. that's #1141

I wouldn't worry about adding any WMS code here. #1141 is probably a long way off, and there's no way to currently test any WMS code you write.

**Comment by bhousel** [comment]
> I wouldn't worry about adding any WMS code here. #1141 is probably a long way off, and there's no way to currently test any WMS code you write.

I'd prefer to leave it in actually.  This isn't WMS code, but rather code for generating an identifier to query the the Offset Database.

**Comment by kepta** [comment]
Hey @Zverik , do you have test cases for this [code](https://trac.openstreetmap.org/browser/subversion/applications/editors/josm/plugins/imagery_offset_db/src/iodb/ImageryIdGenerator.java)

**Comment by bhousel** [comment]
stale

**Pull request closed without merging**